spec/rspec/expectations/block_snippet_extractor_spec.rb

```diff
@@ -0,0 +1,267 @@
+
+      context 'with `target_method(:arg1,:arg2){|arg1,arg2|body}`' do
+        let(:expression) do
+          target_method(:arg1,:arg2){|arg1,arg2|1.positive?}
+        end
+
+        it 'returns the body content lines' do
+          expect(body_content_lines).to eq(['1.positive?'])
+        end
+      end
+
+      context 'with a multiline block containing a single line body' do
+        let(:expression) do
+          target_method do
+            1.positive?
+          end
+        end
+
+        it 'returns the body content lines' do
+          expect(body_content_lines).to eq(['1.positive?'])
+        end
+      end
+
+      context 'with a multiline body block' do
+        let(:expression) do
+          target_method do
+            1.positive?
+            2.negative?
+          end
+        end
+
+        it 'returns the body content lines' do
+          expect(body_content_lines).to eq([
+            '1.positive?',
+            '2.negative?'
+          ])
+        end
+      end
+
+      context 'with `target_method { { :key => "value" } }`' do
+        let(:expression) do
+          target_method { { :key => "value" } }
+        end
+
+        it 'does not confuse the hash curly with the block closer' do
+          expect(body_content_lines).to eq(['{ :key => "value" }'])
+        end
+      end
+
+      context 'with a do-end block containing another do-end block' do
+        let(:expression) do
+          target_method do
+            2.times do |index|
+              puts index
+            end
+          end
+        end
+
+        it 'does not confuse the inner `end` with the outer `end`' do
+          expect(body_content_lines).to eq([
+            '2.times do |index|',
+            'puts index',
+            'end'
+          ])
+        end
+      end
+
+      context "when there's another method invocation on the same line before the target" do
+        let(:expression) do
+          another_method { 2.negative? }; target_method { 1.positive? }
+        end
+
+        it 'correctly extracts the target snippet' do
+          expect(body_content_lines).to eq(['1.positive?'])
+        end
+      end
+
+      context "when there's another method invocation on the same line after the target" do
+        let(:expression) do
+          target_method { 1.positive? }; another_method { 2.negative? }
+        end
+
+        it 'correctly extracts the target snippet' do
+          expect(body_content_lines).to eq(['1.positive?'])
+        end
+      end
+
+      context "when there's another method invocation with the same name on the same line before the target" do
+        let(:expression) do
+          dummy_object.target_method { 2.negative? }; target_method { 1.positive? }
+        end
+
+        let(:dummy_object) do
+          double('dummy_object', :target_method => nil)
+        end
+
+        it 'raises AmbiguousTargetError' do
+          expect { body_content_lines }.to raise_error(BlockSnippetExtractor::AmbiguousTargetError)
```

Comment:
Does this mean the user will get this error instead of a simpler failure message?  That would be bad, IMO.